response/slack/signals.py

```diff
@@ -58,3 +57,114 @@ def update_headline_after_save(sender, instance, **kwargs):
 
     """
     instance.update_in_slack()
+
+
+@receiver(pre_save, sender=Incident)
+def add_timeline_events(sender, instance: Incident, **kwargs):
+    try:
+        prev_state = Incident.objects.get(pk=instance.pk)
+    except Incident.DoesNotExist:
+        # Incident hasn't been saved yet, nothing to do here.
+        return
+
+    if prev_state.lead != instance.lead:
+        update_incident_lead_event(prev_state, instance)
+
+    if prev_state.report != instance.report:
+        update_incident_report_event(prev_state, instance)
+
+    if prev_state.summary != instance.summary:
+        update_incident_summary_event(prev_state, instance)
+
+    if prev_state.impact != instance.impact:
+        update_incident_impact_event(prev_state, instance)
+
+    if prev_state.severity != instance.severity:
+        update_incident_severity_event(prev_state, instance)
+
+def update_incident_lead_event(prev_state, instance):
+    old_lead = None
+    if prev_state.lead:
+        old_lead = ExternalUserSerializer(prev_state.lead).data
+
+    new_lead = None
+    if instance.lead:
+
+        new_lead = ExternalUserSerializer(instance.lead).data
+
+    if prev_state.lead:
+        if instance.lead:
+            text = f"Incident lead changed from {prev_state.lead.display_name} to {instance.lead.display_name}"
+        else:
+            text = f"{prev_state.lead.display_name} was removed as incident lead"
+
+    else:
+        text = f"{instance.lead.display_name} was added as incident lead"
+
+    add_incident_update_event(
+        incident=instance,
+        update_type="incident_lead",
+        text=text,
+        old_value=old_lead,
+        new_value=new_lead,
+    )
+
+
+def update_incident_report_event(prev_state, instance):
+    add_incident_update_event(
+        incident=instance,
+        update_type="incident_report",
+        text=f'Incident report updated from "{prev_state.report}" to "{instance.report}"',
+        old_value=prev_state.report,
+        new_value=instance.report,
+    )
+
+
+def update_incident_summary_event(prev_state, instance):
+    if prev_state.summary:
+        text = (
+            f'Incident summary updated from "{prev_state.summary}" to "{instance.summary}"',
```

Comment:
This update is going to be _huge_ for some incidents.  Do you think it's useful for this to be a timeline event?

Reply:
Just to update - we chatted offline, and decided that it'd be best to push this decision onto consumers of the API on how they want to process this data.